site/content/posts/2023-08-08-v0.25.0-impersonation-proxy-with-external-certs.md

```diff
@@ -0,0 +1,318 @@
+...
+  spec:
+    impersonationProxy:
+      externalEndpoint: impersonation-proxy-mtls.local
+      mode: enabled
+      service:
+        annotations:
+          service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "4000"
+        type: ClusterIP
+      tls:
+        certificateAuthorityData: foobar
+        secretName: impersonation-proxy-tls-serving-cert
+...
+
+# Confirm that the ClusterIP service was automatically created (may take a minute)
+$ kubectl get service pinniped-concierge-impersonation-proxy-cluster-ip \
+    --namespace pinniped-concierge \
+    --output yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+
+# Configure a webhook authenticator to use the local-user-authenticator
+$ cat << EOF > concierge.webhookauthenticator.yaml
+apiVersion: authentication.concierge.pinniped.dev/v1alpha1
+kind: WebhookAuthenticator
+metadata:
+  name: local-user-authenticator
+spec:
+  endpoint: https://local-user-authenticator.local-user-authenticator.svc/authenticate
+  tls:
+    certificateAuthorityData: $(cat local-user-authenticator-ca.pem.b64)
+EOF
+
+# Create the webhook authenticator
+$ kubectl apply \
+    --filename concierge.webhookauthenticator.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Now deploy a Contour `HTTPProxy` ingress that fronts the `ClusterIP` service.
+Note in particular the `spec.tcpproxy` block, which is different than the typical `spec.rules` block. 
+`spec.tcpproxy` is required when using `spec.virtualhost.tls.passthrough: true`. 
+
+See https://projectcontour.io/docs/1.25/config/tls-termination/#tls-session-passthrough for more details.
+
+```shell
+$ cat << EOF > contour-ingress-impersonation-proxy.yaml
+---
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: impersonation-proxy
+  namespace: pinniped-concierge
+spec:
+  virtualhost:
+    fqdn: impersonation-proxy-mtls.local
+    tls:
+      passthrough: true
+  tcpproxy:
+    services:
+    - name: pinniped-concierge-impersonation-proxy-cluster-ip
+      port: 443
+EOF
+
+$ kubectl apply \
+    --filename contour-ingress-impersonation-proxy.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Now generate the Pinniped kubeconfig so that you can perform mTLS with the impersonation proxy.
+
+```shell
+# add 127.0.0.1 impersonation-proxy-mtls.local to your /etc/hosts!
+$ pinniped get kubeconfig \
+    --static-token "pinny:password123" \
+    --concierge-authenticator-type webhook \
+    --concierge-authenticator-name local-user-authenticator \
+    --concierge-mode ImpersonationProxy \
+    --kubeconfig kind-with-contour.kubeconfig.yaml \
+    > pinniped-kubeconfig.yaml
+```
+
+Now perform an action as user pinny!
+
+```shell
+$ kubectl get pods -A \
+    --kubeconfig pinniped-kubeconfig.yaml
+Error from server (Forbidden): pods is forbidden: User "pinny" cannot list resource "pods" in API group "" at the cluster scope: decision made by impersonation-proxy.concierge.pinniped.dev
+```
+
```

Comment:
We could demo the `whoami` command, something like `pinniped whoami --kubeconfig pinniped-kubeconfig.yaml`

Reply:
done, see next push